README.rst
========
Overview
========

.. start-badges

.. list-table::
    :stub-columns: 1


    * - package
      - | |version| |commits-since|
        | |supported-versions| |supported-implementations| |wheel|
        | |license|
    * - quality
      - | |codacy| |codecov|
    * - tests
      - | |travis|

.. |travis| image:: https://api.travis-ci.org/jonathanelscpt/valr-python.svg?branch=master
    :alt: Travis-CI Build Status
    :target: https://travis-ci.org/jonathanelscpt/valr-python

.. |version| image:: https://img.shields.io/pypi/v/valr-python.svg
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/valr-python

.. |wheel| image:: https://img.shields.io/pypi/wheel/valr-python.svg
    :alt: PyPI Wheel
    :target: https://pypi.org/project/valr-python

.. |supported-versions| image:: https://img.shields.io/pypi/pyversions/valr-python.svg
    :alt: Supported versions
    :target: https://pypi.org/project/valr-python

.. |supported-implementations| image:: https://img.shields.io/pypi/implementation/valr-python.svg
    :alt: Supported implementations
    :target: https://pypi.org/project/valr-python

.. |commits-since| image:: https://img.shields.io/github/commits-since/jonathanelscpt/valr-python/v0.2.7.svg
    :alt: Commits since latest release
    :target: https://github.com/jonathanelscpt/valr-python/compare/v0.2.7...master

.. |license| image:: https://img.shields.io/pypi/l/valr-python.svg
    :alt: PyPI License
    :target: https://pypi.org/project/valr-python

.. |codacy| image:: https://api.codacy.com/project/badge/Grade/cb879e2a6be142b88d4e0c2b3a294fb3
    :target: https://www.codacy.com/manual/jonathanelscpt/valr-python?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=jonathanelscpt/valr-python&amp;utm_campaign=Badge_Grade
    :alt: Codacy Code Quality Status

.. |codecov| image:: https://codecov.io/gh/jonathanelscpt/valr-python/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/jonathanelscpt/valr-python
    :alt: Coverage Status


.. end-badges


A Python SDK for VALR cryptocurrency exchange APIs.


Installation
============

::

    pip install valr-python

You can also install the in-development version with::

    pip install https://github.com/jonathanelscpt/valr-python/archive/master.zip



Authentication
==============

Authenticating to VALR API private resources requires a valid API Key from the `VALR exchange <https://www.valr.com/>`_.


Synchronous REST API Client
===========================


To use the **synchronous** REST API client:

.. code-block:: python

    >>> from valr_python import Client
    >>> from valr_python.exceptions import IncompleteOrderWarning
    >>> from decimal import Decimal
    >>>
    >>> c = Client(api_key='api_key', api_secret='api_secret')
    >>> c.rate_limiting_support = True # honour HTTP 429 "Retry-After" header values
    >>> limit_order = {
    ...     "side": "SELL",
    ...     "quantity": Decimal('0.1'),
    ...     "price": Decimal('10000'),
    ...     "pair": "BTCZAR",
    ...     "post_only": True,
    ... }
    >>> try:
    ...    res = c.post_limit_order(**limit_order)
    ...    order_id = res['id']
    ...    print(order_id)
    ... except IncompleteOrderWarning as w:  # HTTP 202 Accepted handling for incomplete orders
    ...    order_id = w.data['id']
    ...    print(order_id)
    ... except Exception as e:
    ...    print(e)
    "558f5e0a-ffd1-46dd-8fae-763d93fa2f25"


Asynchronous REST API Client
============================

The **asynchronous** REST API client is still in development.  *Coming soon!*


WebSocket API Client
====================

To use the WebSocket API client:


.. code-block:: python

    >>> import asyncio
    >>> from typing import Dict
    >>> from pprint import pprint
    >>> from valr_python import WebSocketClient
    >>> from valr_python.enum import TradeEvent
    >>> from valr_python.enum import WebSocketType
    >>>
    >>> def pretty_hook(data: Dict):
    ...    pprint(data)
    >>>
    >>> c = WebSocketClient(api_key='api_key', api_secret='api_secret', currency_pairs=['BTCZAR'],
    ...                     ws_type=WebSocketType.TRADE.name,
    ...                     trade_subscriptions=[TradeEvent.MARKET_SUMMARY_UPDATE.name],
    ...                     hooks={TradeEvent.MARKET_SUMMARY_UPDATE.name: pretty_hook})
    >>> loop = asyncio.get_event_loop()
    >>> loop.run_until_complete(c.run())
    {'currencyPairSymbol': 'BTCZAR',
     'data': {'askPrice': '151601',
              'baseVolume': '314.7631144',
              'bidPrice': '151600',
              'changeFromPrevious': '2.14',
              'created': '2020-02-06T22:47:03.129Z',
              'currencyPairSymbol': 'BTCZAR',
              'highPrice': '152440',
              'lastTradedPrice': '151600',
              'lowPrice': '146765',
              'previousClosePrice': '148410',
              'quoteVolume': '47167382.04552981'},
     'type': 'MARKET_SUMMARY_UPDATE'}


This library leverages :code:`websockets` and :code:`asyncio` and is thus a coroutine-based API client.  Both of
VALR's **Account WebSocket connection** and **Trade WebSocket connection** API endpoints are included.  Furthermore,
the SDK fully supports VALR's subscription methods for both :code:`Account` and :code:`Trade` endpoints.
Please see the `VALR API documentation <https://docs.valr.com/>`_ for further information.

For each subscription, a hook must be provided to process the WS responses.  Failing to do so raises
a :code:`HookNotFoundError` exception.  For ease of use, several :code:`Enum` classes have been implemented
(as showcased above) for client instantiation and hook consumption of API responses. However, client input is
accepted in :code:`str` format.

Although not completely minimalistic, please note that the SDK is implemented as a thin client and parsing of API
streams response is left up to the application user.


Development
===========

To execute all tests run::

    tox

Note, to combine the coverage data from all the tox environments run:

.. list-table::
    :widths: 10 90
    :stub-columns: 1

    - - Windows
      - ::

            set PYTEST_ADDOPTS=--cov-append
            tox

    - - Other
      - ::

            PYTEST_ADDOPTS=--cov-append tox


Donate
======

If this library has helped you or if you would like to support future development, donations are most welcome:

==============  ==========================================
Cryptocurrency  Address
==============  ==========================================
 **BTC**        38c7QWggrB2HLUJZFmhAC2zh4t8C57c1ec
 **ETH**        0x01eD3b58a07c6d005281Db76e6c1AE2bfF2226AD
==============  ==========================================


CHANGELOG.rst

Changelog
=========


0.3.0 (2023-04-20)
------------------

* Added new API features:
* Add fiat bank accounts APIs
* Add VALR Pay APIs
* Added formatting of balances as a convenience function


0.2.7 (2021-12-06)
------------------

* Enum support for SOLZAR


0.2.6 (2021-11-15)
------------------

* added subaccount support
* added all rest api endpoint changes for 2021
* api renaming to align with VALR api docs
* unit tests extended to support endpoint changes and subaccount feature


0.2.5 (2021-11-02)
------------------

* bank fast withdrawal support

0.2.4 (2020-09-27)
------------------

* build stability bugfixes

0.2.3 (2020-09-27)
------------------

* replaced :code:`json` with :code:`simplejson` for speed improvements

0.2.2 (2020-05-25)
------------------

* renamed several sdk methods for API standardization
* added enums for REST API
* added support for simple enum str printing
* added type support for enums
* added XRPZAR support in currency enum

0.2.1 (2020-05-25)
------------------

* :code:`Decimal` support and JSON serialization handling
* added :code:`typing` support for all API calls
* async support for VALR's beta websocket API
* additional api documentation
* build automation extensions

0.1.7 (2020-01-11)
------------------

* Standardised api attribute naming in Client
* Updated doctest and readme for more detailed SDK description

0.1.6 (2020-01-11)
------------------

* Added custom warning :code:`IncompleteOrderWarning` on receiving :code:`202 Accepted` response to support
  custom handling of incomplete orders
* Added custom warning :code:`TooManyRequestsWarning` during 429 handling
* Added class flag for enabling or disabling 429 handler

0.1.5 (2020-01-11)
------------------

* Expanded test suites to unit, functional and live integration testing
* Internal class refactoring
* Added optional HTTP 429 handling by honouring the "Retry-After" header value sent in VALR responses

0.1.4 (2020-01-04)
------------------

* Streamlined ordering api calls with decorators
* Added typing support
* Re-worked class design internals to support later async expansion
* Fixed bugs with str 'None' insertion with f-strings
* Expanded unit tests to cover all api endpoints for synchronous client

0.1.3 (2020-01-02)
------------------

* Fixed defect with empty body signed signatures
* Updated class importing
* Added additional docstrings and unit tests

0.1.2 (2019-12-31)
------------------

* Updated readme and documentation

0.1.1 (2019-12-31)
------------------

* corrected build error issue in setup.py

0.1.0 (2019-12-29)
------------------

* Initial PyPi release

0.0.0 (2019-12-27)
------------------

* Alpha-only. Not released on PyPI.


src/valr_python/client.py
import json
import warnings
from json.decoder import JSONDecodeError
from time import sleep
from typing import Dict
from typing import List
from typing import Optional
from typing import Union

from requests.exceptions import HTTPError

from .base_client import MethodClientABC
from .exceptions import APIException
from .exceptions import IncompleteOrderWarning
from .exceptions import TooManyRequestsWarning


class Client(MethodClientABC):
    """
        Python SDK for the VALR API.

            >>> from valr_python import Client
            >>> from valr_python.exceptions import IncompleteOrderWarning
            >>>
            >>> c = Client(api_key='api_key', api_secret='api_secret')
            >>> c.rate_limiting_support = True # honour HTTP 429 "Retry-After" header values
            >>> limit_order = {
            ...     "side": "SELL",
            ...     "quantity": 0.1,
            ...     "price": 10000,
            ...     "pair": "BTCZAR",
            ...     "post_only": True,
            ... }
            >>> try:
            ...    res = c.post_limit_order(**limit_order)
            ...    order_id = res['id']
            ...    print(order_id)
            ... except IncompleteOrderWarning as w:  # HTTP 202 Accepted handling for incomplete orders
            ...    order_id = w.data['id']
            ...    print(order_id)
            ... except Exception as e:
            ...    print(e)
            "558f5e0a-ffd1-46dd-8fae-763d93fa2f25"
            >>>
        """

    def _do(self, method: str, path: str, data: Dict = None,
            is_authenticated: bool = False) -> Optional[Union[List, Dict]]:
        """Executes API request and returns the response.

        Includes HTTP 429 handling by honouring VALR's 429 Retry-After header cool-down.

        :param method: HTTP method (e.g. GET, POST, DELETE, etc.
        :param path: REST API endpoint path
        :param is_authenticated: True if api call requires authentication and signature hashing
        :param data: request data
        :return: HTTP REST API response from VALR
        """
        params = json.loads(json.dumps(data))
        headers = {}
        url = self._base_url + '/' + path.lstrip('/')
        if is_authenticated:
            headers.update(self._get_valr_headers(method=method, path=path, params=params))
        if data:
            headers["Content-Type"] = "application/json"
        args = dict(timeout=self._timeout, params=params, headers=headers)
        res = self._session.request(method, url, **args)

        try:
            res.raise_for_status()
            e = res.json()
            self._raise_for_api_error(e)
            # provide warning with bundled response dict for incomplete transactions
            if res.status_code == 202:
                warnings.warn(IncompleteOrderWarning(data=e, message="Order processing incomplete"))
            return e
        except HTTPError as he:
            if res.status_code == 429:
                if self._rate_limiting_support:
                    try:
                        retry_after = float(res.headers['Retry-After'])
                        warnings.warn(f"HTTP 429 response received. Applying Retry-After {retry_after}sec back-off",
                                      TooManyRequestsWarning)
                        sleep(retry_after)
                        return self._do(method=method, path=path, is_authenticated=is_authenticated, data=data)
                    except (KeyError, ValueError):
                        raise APIException(res.status_code,
                                           f'valr-python: HTTP 429 processing failed. '
                                           f'HTTP ({res.status_code}): {res.headers}')
                else:
                    # avoid JSONDecodeError - VALR 429 response has html body
                    raise he
            e = res.json()
            self._raise_for_api_error(e)
            # bubble HTTP errors that VALR API doesn't report on
            raise he
        except JSONDecodeError as jde:
            raise APIException(res.status_code,
                               f'valr-python: unknown API error. HTTP ({res.status_code}): {jde.msg}')
